rust/relay/src/main.rs

```diff
@@ -19,36 +26,120 @@ async fn main() -> Result<()> {
+    let listen_ip4_addr = parse_env_var::<Ipv4Addr>("RELAY_LISTEN_IP4_ADDR")?;
+
+    let socket = UdpSocket::bind((listen_ip4_addr, 3478)).await?;
+    let mut recv_buf = [0u8; MAX_UDP_SIZE];
 
-    tracing::info!("Listening on: {addr}", addr = socket.local_addr()?);
+    let mut server = Server::new(SocketAddrV4::new(public_ip4_addr, 3478));
 
-    let mut server = Server::default();
+    tracing::info!("Listening for incoming traffic on UDP port 3478");
 
-    let mut buf = [0u8; MAX_UDP_SIZE];
+    let mut wake = pin!(Sleep::default());
+    let (relayed_data_sender, mut relayed_data_receiver) =
+        futures::channel::mpsc::channel::<(Vec<u8>, SocketAddr, AllocationId)>(10);
+    let mut allocation_tasks = HashMap::<AllocationId, tokio::task::JoinHandle<()>>::default();
 
     loop {
-        // TODO: Listen for websocket commands here and update the server state accordingly.
-        let (recv_len, sender) = socket.recv_from(&mut buf).await?;
+        tokio::select! {
+            () = &mut wake => {
+                server.handle_deadline_reached(Instant::now());
+            }
+            (payload, sender, allocation_id) = relayed_data_receiver.select_next_some() => {
+                if tracing::enabled!(target: "wire", Level::TRACE) {
+                    let hex_bytes = hex::encode(&payload);
+                    tracing::trace!(target: "wire", r#"Input::Relay("{sender}","{hex_bytes}")"#);
+                }
 
-        if tracing::enabled!(target: "wire", Level::TRACE) {
-            let hex_bytes = hex::encode(&buf[..recv_len]);
-            tracing::trace!(target: "wire", r#"Input("{sender}","{}")"#, hex_bytes);
-        }
+                server.handle_relay_input(&payload, sender, allocation_id);
+            }
+            receive_result = pin!(socket.recv_from(&mut recv_buf)) => {
+                let (length, sender) = receive_result.context("Failed to receive from socket")?;
+                let payload = &recv_buf[..length];
 
-        match server.handle_received_bytes(&buf[..recv_len], sender) {
-            Ok(Some((response, recipient))) => {
                 if tracing::enabled!(target: "wire", Level::TRACE) {
-                    let hex_bytes = hex::encode(&response);
-                    tracing::trace!(target: "wire", r#"Output("{recipient}","{}")"#, hex_bytes);
+                    let hex_bytes = hex::encode(payload);
+                    tracing::trace!(target: "wire", r#"Input::Client("{sender}","{hex_bytes}")"#);
                 }
 
-                socket.send_to(&response, recipient).await?;
+                if let Err(e) = server.handle_client_input(payload, sender, Instant::now()) {
+                    tracing::debug!("Failed to handle datagram from {sender}: {e}")
+                }
             }
-            Ok(None) => {}
-            Err(e) => {
-                tracing::debug!("Failed to handle datagram from {sender}: {e}")
+        }
```

Comment:
I am not a big fan of `tokio::select!` because it has a custom syntax and subtle behaviours regarding polling fairness. I plan to rewrite this into a state machine that just has a `poll` function and polls these things in a clear (prioritized) order. For now, this will do though.